{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Accessing 10x RNA-seq gene expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import anndata\n",
    "\n",
    "from abc_atlas_access.abc_atlas_cache.abc_project_cache import AbcProjectCache\n",
    "from abc_atlas_access.abc_atlas_cache.anndata_utils import get_gene_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will interact with the data using the **AbcProjectCache**. \n",
    "\n",
    "**Change the download_base variable to where you have downloaded the data in your system.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\guoyu\\anaconda3\\envs\\allen\\Lib\\site-packages\\abc_atlas_access\\abc_atlas_cache\\cloud_cache.py:630: OutdatedManifestWarning: You are loading releases/20250531/manifest.json. A more up to date version of the dataset -- releases/20251031/manifest.json -- exists online. To see the changes between the two versions of the dataset, run\n",
      "type.compare_manifests('releases/20250531/manifest.json', 'releases/20251031/manifest.json')\n",
      "To load another version of the dataset, run\n",
      "type.load_manifest('releases/20251031/manifest.json')\n",
      "  warnings.warn(msg, OutdatedManifestWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'releases/20250531/manifest.json'"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "download_base = Path('../../data/abc_atlas')\n",
    "abc_cache = AbcProjectCache.from_cache_dir(download_base)\n",
    "\n",
    "abc_cache.current_manifest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "type object 'type' has no attribute 'load_manifest'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[117], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28mtype\u001b[39m\u001b[38;5;241m.\u001b[39mload_manifest(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mreleases/20251031/manifest.json\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[1;31mAttributeError\u001b[0m: type object 'type' has no attribute 'load_manifest'"
     ]
    }
   ],
   "source": [
    "type.load_manifest('releases/20251031/manifest.json')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>gene_count</th>\n",
       "      <th>umi_count</th>\n",
       "      <th>doublet_score</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>cluster_alias</th>\n",
       "      <th>cell_in_wmb_study</th>\n",
       "      <th>wmb_cluster_alias</th>\n",
       "      <th>library_label</th>\n",
       "      <th>...</th>\n",
       "      <th>proportion_adult_cells</th>\n",
       "      <th>proportion_aged_cells</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>log2_odds_ratio</th>\n",
       "      <th>cluster_age_bias</th>\n",
       "      <th>max_region_of_interest_color</th>\n",
       "      <th>cluster_age_bias_color</th>\n",
       "      <th>neurotransmitter_combined_label</th>\n",
       "      <th>neurotransmitter_label</th>\n",
       "      <th>neurotransmitter_color</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCCTGTTGTGAATTAG-135_B01</th>\n",
       "      <td>GCCTGTTGTGAATTAG</td>\n",
       "      <td>6777</td>\n",
       "      <td>37834.0</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>-1.185296</td>\n",
       "      <td>1.881103</td>\n",
       "      <td>278</td>\n",
       "      <td>True</td>\n",
       "      <td>1079</td>\n",
       "      <td>L8TX_190716_01_D07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.577949</td>\n",
       "      <td>0.422051</td>\n",
       "      <td>1.013716</td>\n",
       "      <td>0.019653</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>#80CDF8</td>\n",
       "      <td>#DADEDF</td>\n",
       "      <td>GABA</td>\n",
       "      <td>GABA</td>\n",
       "      <td>#FF3358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCGAAAGTGAAGCGT-761_A04</th>\n",
       "      <td>TCCGAAAGTGAAGCGT</td>\n",
       "      <td>3396</td>\n",
       "      <td>9585.0</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>14.007499</td>\n",
       "      <td>-0.220336</td>\n",
       "      <td>817</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_210805_01_H01</td>\n",
       "      <td>...</td>\n",
       "      <td>0.454494</td>\n",
       "      <td>0.545506</td>\n",
       "      <td>1.046448</td>\n",
       "      <td>0.065500</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>#8599CC</td>\n",
       "      <td>#DADEDF</td>\n",
       "      <td>No-NT</td>\n",
       "      <td>No-NT</td>\n",
       "      <td>#666666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTACCATGTCGTGGTC-327_A06</th>\n",
       "      <td>TTACCATGTCGTGGTC</td>\n",
       "      <td>4294</td>\n",
       "      <td>10527.0</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>6.550781</td>\n",
       "      <td>-1.378886</td>\n",
       "      <td>804</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_200813_01_H10</td>\n",
       "      <td>...</td>\n",
       "      <td>0.370570</td>\n",
       "      <td>0.629430</td>\n",
       "      <td>1.741208</td>\n",
       "      <td>0.800089</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>#80C0E2</td>\n",
       "      <td>#DADEDF</td>\n",
       "      <td>No-NT</td>\n",
       "      <td>No-NT</td>\n",
       "      <td>#666666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACGGGTCGTACGAGCA-385_D06</th>\n",
       "      <td>ACGGGTCGTACGAGCA</td>\n",
       "      <td>2943</td>\n",
       "      <td>6957.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.650183</td>\n",
       "      <td>0.289086</td>\n",
       "      <td>817</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_201008_01_A12</td>\n",
       "      <td>...</td>\n",
       "      <td>0.454494</td>\n",
       "      <td>0.545506</td>\n",
       "      <td>1.046448</td>\n",
       "      <td>0.065500</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>#8599CC</td>\n",
       "      <td>#DADEDF</td>\n",
       "      <td>No-NT</td>\n",
       "      <td>No-NT</td>\n",
       "      <td>#666666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTAGGATCTGTCCCA-301_B04</th>\n",
       "      <td>CTTAGGATCTGTCCCA</td>\n",
       "      <td>7907</td>\n",
       "      <td>47406.0</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>6.583878</td>\n",
       "      <td>-10.639808</td>\n",
       "      <td>152</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_200723_01_B10</td>\n",
       "      <td>...</td>\n",
       "      <td>0.403409</td>\n",
       "      <td>0.596591</td>\n",
       "      <td>0.587232</td>\n",
       "      <td>-0.767999</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>#72D569</td>\n",
       "      <td>#DADEDF</td>\n",
       "      <td>Glut</td>\n",
       "      <td>Glut</td>\n",
       "      <td>#2B93DF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              cell_barcode  gene_count  umi_count  \\\n",
       "cell_label                                                          \n",
       "GCCTGTTGTGAATTAG-135_B01  GCCTGTTGTGAATTAG        6777    37834.0   \n",
       "TCCGAAAGTGAAGCGT-761_A04  TCCGAAAGTGAAGCGT        3396     9585.0   \n",
       "TTACCATGTCGTGGTC-327_A06  TTACCATGTCGTGGTC        4294    10527.0   \n",
       "ACGGGTCGTACGAGCA-385_D06  ACGGGTCGTACGAGCA        2943     6957.0   \n",
       "CTTAGGATCTGTCCCA-301_B04  CTTAGGATCTGTCCCA        7907    47406.0   \n",
       "\n",
       "                          doublet_score          x          y  cluster_alias  \\\n",
       "cell_label                                                                     \n",
       "GCCTGTTGTGAATTAG-135_B01       0.200000  -1.185296   1.881103            278   \n",
       "TCCGAAAGTGAAGCGT-761_A04       0.030303  14.007499  -0.220336            817   \n",
       "TTACCATGTCGTGGTC-327_A06       0.020000   6.550781  -1.378886            804   \n",
       "ACGGGTCGTACGAGCA-385_D06       0.000000  13.650183   0.289086            817   \n",
       "CTTAGGATCTGTCCCA-301_B04       0.037037   6.583878 -10.639808            152   \n",
       "\n",
       "                          cell_in_wmb_study  wmb_cluster_alias  \\\n",
       "cell_label                                                       \n",
       "GCCTGTTGTGAATTAG-135_B01               True               1079   \n",
       "TCCGAAAGTGAAGCGT-761_A04              False               <NA>   \n",
       "TTACCATGTCGTGGTC-327_A06              False               <NA>   \n",
       "ACGGGTCGTACGAGCA-385_D06              False               <NA>   \n",
       "CTTAGGATCTGTCCCA-301_B04              False               <NA>   \n",
       "\n",
       "                               library_label  ...  proportion_adult_cells  \\\n",
       "cell_label                                    ...                           \n",
       "GCCTGTTGTGAATTAG-135_B01  L8TX_190716_01_D07  ...                0.577949   \n",
       "TCCGAAAGTGAAGCGT-761_A04  L8TX_210805_01_H01  ...                0.454494   \n",
       "TTACCATGTCGTGGTC-327_A06  L8TX_200813_01_H10  ...                0.370570   \n",
       "ACGGGTCGTACGAGCA-385_D06  L8TX_201008_01_A12  ...                0.454494   \n",
       "CTTAGGATCTGTCCCA-301_B04  L8TX_200723_01_B10  ...                0.403409   \n",
       "\n",
       "                         proportion_aged_cells odds_ratio log2_odds_ratio  \\\n",
       "cell_label                                                                  \n",
       "GCCTGTTGTGAATTAG-135_B01              0.422051   1.013716        0.019653   \n",
       "TCCGAAAGTGAAGCGT-761_A04              0.545506   1.046448        0.065500   \n",
       "TTACCATGTCGTGGTC-327_A06              0.629430   1.741208        0.800089   \n",
       "ACGGGTCGTACGAGCA-385_D06              0.545506   1.046448        0.065500   \n",
       "CTTAGGATCTGTCCCA-301_B04              0.596591   0.587232       -0.767999   \n",
       "\n",
       "                         cluster_age_bias max_region_of_interest_color  \\\n",
       "cell_label                                                               \n",
       "GCCTGTTGTGAATTAG-135_B01       unassigned                      #80CDF8   \n",
       "TCCGAAAGTGAAGCGT-761_A04       unassigned                      #8599CC   \n",
       "TTACCATGTCGTGGTC-327_A06       unassigned                      #80C0E2   \n",
       "ACGGGTCGTACGAGCA-385_D06       unassigned                      #8599CC   \n",
       "CTTAGGATCTGTCCCA-301_B04       unassigned                      #72D569   \n",
       "\n",
       "                          cluster_age_bias_color  \\\n",
       "cell_label                                         \n",
       "GCCTGTTGTGAATTAG-135_B01                 #DADEDF   \n",
       "TCCGAAAGTGAAGCGT-761_A04                 #DADEDF   \n",
       "TTACCATGTCGTGGTC-327_A06                 #DADEDF   \n",
       "ACGGGTCGTACGAGCA-385_D06                 #DADEDF   \n",
       "CTTAGGATCTGTCCCA-301_B04                 #DADEDF   \n",
       "\n",
       "                         neurotransmitter_combined_label  \\\n",
       "cell_label                                                 \n",
       "GCCTGTTGTGAATTAG-135_B01                            GABA   \n",
       "TCCGAAAGTGAAGCGT-761_A04                           No-NT   \n",
       "TTACCATGTCGTGGTC-327_A06                           No-NT   \n",
       "ACGGGTCGTACGAGCA-385_D06                           No-NT   \n",
       "CTTAGGATCTGTCCCA-301_B04                            Glut   \n",
       "\n",
       "                         neurotransmitter_label neurotransmitter_color  \n",
       "cell_label                                                              \n",
       "GCCTGTTGTGAATTAG-135_B01                   GABA                #FF3358  \n",
       "TCCGAAAGTGAAGCGT-761_A04                  No-NT                #666666  \n",
       "TTACCATGTCGTGGTC-327_A06                  No-NT                #666666  \n",
       "ACGGGTCGTACGAGCA-385_D06                  No-NT                #666666  \n",
       "CTTAGGATCTGTCCCA-301_B04                   Glut                #2B93DF  \n",
       "\n",
       "[5 rows x 60 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the cell metadata.\n",
    "cell = abc_cache.get_metadata_dataframe(\n",
    "    directory='Zeng-Aging-Mouse-10Xv3',\n",
    "    file_name='cell_metadata',\n",
    "    dtype={'cell_label': str,\n",
    "           'wmb_cluster_alias': 'Int64'}\n",
    ")\n",
    "cell.set_index('cell_label', inplace=True)\n",
    "\n",
    "cell_colors = abc_cache.get_metadata_dataframe(\n",
    "    directory='Zeng-Aging-Mouse-10Xv3',\n",
    "    file_name='cell_annotation_colors'\n",
    ").set_index('cell_label')\n",
    "\n",
    "cluster_info = abc_cache.get_metadata_dataframe(\n",
    "    directory='Zeng-Aging-Mouse-10Xv3',\n",
    "    file_name='cluster'\n",
    ").set_index('cluster_alias')\n",
    "\n",
    "cell_cluster_mapping = abc_cache.get_metadata_dataframe(\n",
    "    directory='Zeng-Aging-Mouse-WMB-taxonomy',\n",
    "    file_name='cell_cluster_mapping_annotations'\n",
    ").set_index('cell_label')\n",
    "cell_cluster_mapping.head()\n",
    "\n",
    "# Join on the cell_label index.\n",
    "cell_extended = cell.join(cell_cluster_mapping, rsuffix='_cl_map')\n",
    "cell_extended = cell_extended.join(cell_colors, rsuffix='_cl_colors')\n",
    "# Join the cluster information in by joining on the Aging dataset's cluster_alias column.\n",
    "cell_extended = cell_extended.join(cluster_info, on='cluster_alias', rsuffix='_cl_info')\n",
    "\n",
    "# Quick run through to drop duplicated columns\n",
    "drop_cols = []\n",
    "for col in cell_extended.columns:\n",
    "    if col.endswith(('_cl_map', '_cl_colors', '_cl_info')):\n",
    "        drop_cols.append(col)\n",
    "cell_extended.drop(columns=drop_cols, inplace=True)\n",
    "\n",
    "# The dataset is sorted on cell_label by default, this causes some plotting weirdness\n",
    "# due to all \"adult\" cells being first in the order. Below we scrabble the DataFrame\n",
    "# to better reproduce plots from the paper.\n",
    "cell_extended = cell_extended.sample(frac=1, random_state=12345)\n",
    "\n",
    "# del cell\n",
    "del cell_colors\n",
    "del cell_cluster_mapping\n",
    "del cluster_info\n",
    "\n",
    "cell_extended.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Column names in cell: ['cell_barcode', 'gene_count', 'umi_count', 'doublet_score', 'x', 'y', 'cluster_alias', 'cell_in_wmb_study', 'wmb_cluster_alias', 'library_label', 'alignment_job_id', 'library_method', 'barcoded_cell_sample_label', 'enrichment_population', 'region_of_interest_label', 'anatomical_division_label', 'library_in_wmb_study', 'donor_label', 'population_sampling', 'donor_genotype', 'donor_sex', 'donor_age', 'donor_age_category', 'donor_in_wmb_study', 'feature_matrix_label', 'dataset_label', 'abc_sample_id', 'cluster_label', 'cluster_order', 'cluster_name', 'cluster_color', 'class_name', 'subclass_name', 'supertype_name', 'class_color', 'subclass_color', 'supertype_color', 'anatomical_division_color', 'anatomical_division_order', 'donor_age_category_color', 'donor_age_category_order', 'donor_sex_color', 'donor_sex_order', 'region_of_interest_color', 'region_of_interest_order', 'number_of_cells', 'max_region_of_interest_label', 'proportion_max_region_of_interest_label', 'number_of_adult_cells', 'number_of_aged_cells', 'proportion_adult_cells', 'proportion_aged_cells', 'odds_ratio', 'log2_odds_ratio', 'cluster_age_bias', 'max_region_of_interest_color', 'cluster_age_bias_color', 'neurotransmitter_combined_label', 'neurotransmitter_label', 'neurotransmitter_color']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class_name\n",
       "30 Astro-Epen        266372\n",
       "31 OPC-Oligo         253468\n",
       "01 IT-ET Glut        166299\n",
       "33 Vascular           84721\n",
       "34 Immune             72736\n",
       "02 NP-CT-L6b Glut     58901\n",
       "19 MB Glut            48744\n",
       "09 CNU-LGE GABA       42802\n",
       "20 MB GABA            23894\n",
       "26 P GABA             23821\n",
       "06 CTX-CGE GABA       19256\n",
       "07 CTX-MGE GABA       19232\n",
       "23 P Glut             17158\n",
       "12 HY GABA            15960\n",
       "14 HY Glut            14508\n",
       "04 DG-IMN Glut        11984\n",
       "21 MB Dopa             5544\n",
       "11 CNU-HYa GABA        3735\n",
       "16 HY MM Glut          3345\n",
       "22 MB-HB Sero          2898\n",
       "08 CNU-MGE GABA        2126\n",
       "03 OB-CR Glut          1761\n",
       "13 CNU-HYa Glut        1399\n",
       "24 MY Glut             1077\n",
       "05 OB-IMN GABA          824\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Column names in cell:', cell_extended.columns.to_list())\n",
    "cell_extended['class_name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "region_of_interest_label\n",
       "MB - PAG-RAmb                  143187\n",
       "HPF - ENT                       89730\n",
       "HY - HY                         82431\n",
       "HPF - HIP                       78199\n",
       "Iscortex  - ACA                 76283\n",
       "P - Pmot/sat-post               70309\n",
       "Isocortex  - AI                 70288\n",
       "P - Pmot/sat-ant                70031\n",
       "Isocortex - RSP                 69541\n",
       "MB - VTA-SN                     69333\n",
       "HPF - PAR-POST-PRE-SUB-ProS     69056\n",
       "PAL - PAL                       66139\n",
       "Isocortex - PL-ILA-ORB          64690\n",
       "STR - STRv                      58034\n",
       "STR - sAMY                      55600\n",
       "STR - STRd                      29714\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_extended['region_of_interest_label'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>mapped_ncbi_identifier</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000035131</th>\n",
       "      <td>Brinp3</td>\n",
       "      <td>bone morphogenetic protein/retinoic acid induc...</td>\n",
       "      <td>NCBIGene:215378</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000026697</th>\n",
       "      <td>Myoc</td>\n",
       "      <td>myocilin</td>\n",
       "      <td>NCBIGene:17926</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000031548</th>\n",
       "      <td>Sfrp1</td>\n",
       "      <td>secreted frizzled-related protein 1</td>\n",
       "      <td>NCBIGene:20377</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000032087</th>\n",
       "      <td>Dscaml1</td>\n",
       "      <td>DS cell adhesion molecule like 1</td>\n",
       "      <td>NCBIGene:114873</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000076431</th>\n",
       "      <td>Sox4</td>\n",
       "      <td>SRY (sex determining region Y)-box 4</td>\n",
       "      <td>NCBIGene:20677</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000069170</th>\n",
       "      <td>Adgrv1</td>\n",
       "      <td>adhesion G protein-coupled receptor V1</td>\n",
       "      <td>NCBIGene:110789</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000091722</th>\n",
       "      <td>Siah3</td>\n",
       "      <td>siah E3 ubiquitin protein ligase family member 3</td>\n",
       "      <td>NCBIGene:380918</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000022112</th>\n",
       "      <td>Gpc5</td>\n",
       "      <td>glypican 5</td>\n",
       "      <td>NCBIGene:103978</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000059213</th>\n",
       "      <td>Ddn</td>\n",
       "      <td>dendrin</td>\n",
       "      <td>NCBIGene:13199</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000024411</th>\n",
       "      <td>Aqp4</td>\n",
       "      <td>aquaporin 4</td>\n",
       "      <td>NCBIGene:11829</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000043969</th>\n",
       "      <td>Emx2</td>\n",
       "      <td>empty spiracles homeobox 2</td>\n",
       "      <td>NCBIGene:13797</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_symbol  \\\n",
       "gene_identifier                  \n",
       "ENSMUSG00000035131      Brinp3   \n",
       "ENSMUSG00000026697        Myoc   \n",
       "ENSMUSG00000031548       Sfrp1   \n",
       "ENSMUSG00000032087     Dscaml1   \n",
       "ENSMUSG00000076431        Sox4   \n",
       "ENSMUSG00000069170      Adgrv1   \n",
       "ENSMUSG00000091722       Siah3   \n",
       "ENSMUSG00000022112        Gpc5   \n",
       "ENSMUSG00000059213         Ddn   \n",
       "ENSMUSG00000024411        Aqp4   \n",
       "ENSMUSG00000043969        Emx2   \n",
       "\n",
       "                                                                 name  \\\n",
       "gene_identifier                                                         \n",
       "ENSMUSG00000035131  bone morphogenetic protein/retinoic acid induc...   \n",
       "ENSMUSG00000026697                                           myocilin   \n",
       "ENSMUSG00000031548                secreted frizzled-related protein 1   \n",
       "ENSMUSG00000032087                   DS cell adhesion molecule like 1   \n",
       "ENSMUSG00000076431               SRY (sex determining region Y)-box 4   \n",
       "ENSMUSG00000069170             adhesion G protein-coupled receptor V1   \n",
       "ENSMUSG00000091722   siah E3 ubiquitin protein ligase family member 3   \n",
       "ENSMUSG00000022112                                         glypican 5   \n",
       "ENSMUSG00000059213                                            dendrin   \n",
       "ENSMUSG00000024411                                        aquaporin 4   \n",
       "ENSMUSG00000043969                         empty spiracles homeobox 2   \n",
       "\n",
       "                   mapped_ncbi_identifier comment  \n",
       "gene_identifier                                    \n",
       "ENSMUSG00000035131        NCBIGene:215378     NaN  \n",
       "ENSMUSG00000026697         NCBIGene:17926     NaN  \n",
       "ENSMUSG00000031548         NCBIGene:20377     NaN  \n",
       "ENSMUSG00000032087        NCBIGene:114873     NaN  \n",
       "ENSMUSG00000076431         NCBIGene:20677     NaN  \n",
       "ENSMUSG00000069170        NCBIGene:110789     NaN  \n",
       "ENSMUSG00000091722        NCBIGene:380918     NaN  \n",
       "ENSMUSG00000022112        NCBIGene:103978     NaN  \n",
       "ENSMUSG00000059213         NCBIGene:13199     NaN  \n",
       "ENSMUSG00000024411         NCBIGene:11829     NaN  \n",
       "ENSMUSG00000043969         NCBIGene:13797     NaN  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get gene name and symbol information\n",
    "\n",
    "# select only a few genes\n",
    "gene_names = ['Aqp4', 'Gpc5', 'Dscaml1', 'Sox4', 'Myoc', 'Siah3', 'Adgrv1', 'Emx2', 'Brinp3', 'Ddn', 'Sfrp1']\n",
    "gene = abc_cache.get_metadata_dataframe(directory='WMB-10X', file_name='gene').set_index('gene_identifier')\n",
    "gene.query(\"gene_symbol in @gene_names\")\n",
    "# gene[gene['gene_symbol'].isin(gene_names)]\n",
    "# gene.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading specific genes from the data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading file: Zeng-Aging-Mouse-10Xv3\n"
     ]
    }
   ],
   "source": [
    "# define genes\n",
    "gene_names = ['Aqp4', 'Gpc5', 'Dscaml1', 'Sox4', 'Myoc', 'Siah3', 'Adgrv1', 'Emx2', 'Brinp3', 'Ddn', 'Sfrp1']\n",
    "\n",
    "# define cell subgroup\n",
    "hippo_cells = cell_extended[cell_extended['region_of_interest_label'] == 'HPF - HIP']\n",
    "cortex_cells = cell_extended.query(\"region_of_interest_label  == 'Iscortex  - ACA' & class_name ==  '30 Astro-Epen'\")\n",
    "example_cell = cell_extended[cell_extended['cell_barcode'] == 'AAACCCAAGAATAACC'] # one example cell\n",
    "\n",
    "# read the expression matrix\n",
    "gene_data = get_gene_data(\n",
    "    abc_atlas_cache=abc_cache,\n",
    "    all_cells=cortex_cells,\n",
    "    all_genes=gene,\n",
    "    selected_genes=gene_names\n",
    ")\n",
    "gene_data[pd.notna(gene_data[gene_data.columns[0]])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\u001b[1;31mSignature:\u001b[0m\n",
       "\u001b[0mget_gene_data\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m\n",
       "\u001b[0m    \u001b[0mabc_atlas_cache\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mabc_atlas_access\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mabc_atlas_cache\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mabc_project_cache\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mAbcProjectCache\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\n",
       "\u001b[0m    \u001b[0mall_cells\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mpandas\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcore\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mframe\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\n",
       "\u001b[0m    \u001b[0mall_genes\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mpandas\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcore\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mframe\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\n",
       "\u001b[0m    \u001b[0mselected_genes\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mList\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\n",
       "\u001b[0m    \u001b[0mdata_type\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m'log2'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\n",
       "\u001b[0m    \u001b[0mchunk_size\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mint\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;36m8192\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\n",
       "\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
       "\u001b[1;31mDocstring:\u001b[0m\n",
       "Load expression matrix data from the ABC Atlas and extract data for\n",
       "specific genes.\n",
       "\n",
       "Method will load all expression data required to process across multiple\n",
       "files to extract the full set of genes. This may result in downloading\n",
       "potentially ~100 GB of data.\n",
       "\n",
       "Parameters\n",
       "----------\n",
       "abc_atlas_cache: AbcProjectCache\n",
       "    An AbcProjectCache instance object to handle downloading and serving\n",
       "    the path to the expression matrix data.\n",
       "all_cells: pandas.DataFrame\n",
       "    cells metadata loaded as a pandas Dataframe from the AbcProjectCache\n",
       "    indexed on cell_label.\n",
       "all_genes: pandas.DataFrame\n",
       "    genes metadata loaded as a pandas Dataframe from the AbcProjectCache\n",
       "    indexed on gene_identifier.\n",
       "selected_genes: list of strings\n",
       "    List of gene_symbols that are a subset of those in the full genes\n",
       "    DataFrame.\n",
       "data_type: str (Default: \"log2\")\n",
       "    Kind of expression matrix to load either \"log2\" or \"raw\". Defaults to\n",
       "    \"log2\".\n",
       "chunk_size: int (Default: 8192)\n",
       "    Size of the chunk to load from the anndata files. Adjust this size if\n",
       "    needed based on memory/file io. Default: 8192.\n",
       "\n",
       "Returns\n",
       "-------\n",
       "output_gene_data: pandas.DataFrame\n",
       "    Subset of gene data indexed by cell.\n",
       "\u001b[1;31mFile:\u001b[0m      c:\\users\\guoyu\\anaconda3\\envs\\allen\\lib\\site-packages\\abc_atlas_access\\abc_atlas_cache\\anndata_utils.py\n",
       "\u001b[1;31mType:\u001b[0m      function"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?get_gene_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_symbol</th>\n",
       "      <th>Brinp3</th>\n",
       "      <th>Myoc</th>\n",
       "      <th>Sfrp1</th>\n",
       "      <th>Dscaml1</th>\n",
       "      <th>Sox4</th>\n",
       "      <th>Adgrv1</th>\n",
       "      <th>Siah3</th>\n",
       "      <th>Gpc5</th>\n",
       "      <th>Ddn</th>\n",
       "      <th>Aqp4</th>\n",
       "      <th>Emx2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTACAGTGTATCGCTA-310_A05</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.137723</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.44112</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGGTAGGGTCTCCTGT-316_C03</th>\n",
       "      <td>9.503889</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.517732</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.961513</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.509838</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGATGGTAGCTACTAC-393_C05</th>\n",
       "      <td>9.009558</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.984214</td>\n",
       "      <td>7.017931</td>\n",
       "      <td>9.330926</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCGATTAGATGCTTC-316_A03</th>\n",
       "      <td>8.81147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.978538</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.814678</td>\n",
       "      <td>6.821072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTGGATCCTTCACG-314_B01</th>\n",
       "      <td>9.674965</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0988</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.410429</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0988</td>\n",
       "      <td>8.093528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGATGCTTCTCGTTTA-310_A05</th>\n",
       "      <td>10.977281</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.81947</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.361503</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.809857</td>\n",
       "      <td>6.81947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACTTGTAGCTACAAA-393_C05</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.29075</td>\n",
       "      <td>7.837065</td>\n",
       "      <td>11.079336</td>\n",
       "      <td>7.837065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAACCTCAGCTCGTGC-352_A04</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.300072</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.132433</td>\n",
       "      <td>8.548756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGATCCAAGACCTCCG-393_D05</th>\n",
       "      <td>8.950899</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.371756</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.021403</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.371756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGATCTAGTTGAATG-316_C03</th>\n",
       "      <td>9.750643</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.17403</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.071063</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.17403</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12334 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_symbol                  Brinp3 Myoc    Sfrp1   Dscaml1      Sox4 Adgrv1  \\\n",
       "cell_label                                                                     \n",
       "GTACAGTGTATCGCTA-310_A05        0.0  0.0      0.0       0.0       0.0    0.0   \n",
       "AGGTAGGGTCTCCTGT-316_C03   9.503889  0.0      0.0  6.517732       0.0    0.0   \n",
       "TGATGGTAGCTACTAC-393_C05   9.009558  0.0      0.0       0.0       0.0    0.0   \n",
       "TTCGATTAGATGCTTC-316_A03    8.81147  0.0      0.0       0.0       0.0    0.0   \n",
       "TTGTGGATCCTTCACG-314_B01   9.674965  0.0      0.0       0.0    7.0988    0.0   \n",
       "...                             ...  ...      ...       ...       ...    ...   \n",
       "AGATGCTTCTCGTTTA-310_A05  10.977281  0.0      0.0       0.0   6.81947    0.0   \n",
       "TACTTGTAGCTACAAA-393_C05        0.0  0.0      0.0       0.0       0.0    0.0   \n",
       "CAACCTCAGCTCGTGC-352_A04        0.0  0.0      0.0       0.0       0.0    0.0   \n",
       "AGATCCAAGACCTCCG-393_D05   8.950899  0.0      0.0       0.0  7.371756    0.0   \n",
       "CCGATCTAGTTGAATG-316_C03   9.750643  0.0  7.17403       0.0       0.0    0.0   \n",
       "\n",
       "gene_symbol              Siah3       Gpc5       Ddn       Aqp4      Emx2  \n",
       "cell_label                                                                \n",
       "GTACAGTGTATCGCTA-310_A05   0.0  12.137723       0.0    8.44112       0.0  \n",
       "AGGTAGGGTCTCCTGT-316_C03   0.0  12.961513       0.0   7.509838       0.0  \n",
       "TGATGGTAGCTACTAC-393_C05   0.0  12.984214  7.017931   9.330926       0.0  \n",
       "TTCGATTAGATGCTTC-316_A03   0.0  11.978538       0.0   7.814678  6.821072  \n",
       "TTGTGGATCCTTCACG-314_B01   0.0  12.410429       0.0     7.0988  8.093528  \n",
       "...                        ...        ...       ...        ...       ...  \n",
       "AGATGCTTCTCGTTTA-310_A05   0.0  12.361503       0.0   8.809857   6.81947  \n",
       "TACTTGTAGCTACAAA-393_C05   0.0   11.29075  7.837065  11.079336  7.837065  \n",
       "CAACCTCAGCTCGTGC-352_A04   0.0  12.300072       0.0   9.132433  8.548756  \n",
       "AGATCCAAGACCTCCG-393_D05   0.0  13.021403       0.0        0.0  7.371756  \n",
       "CCGATCTAGTTGAATG-316_C03   0.0  13.071063       0.0    7.17403       0.0  \n",
       "\n",
       "[12334 rows x 11 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The returned DataFrame is indexed by ``cell_label`` and can thus be joined with the ``cell`` DataFrame for further analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>gene_count</th>\n",
       "      <th>umi_count</th>\n",
       "      <th>doublet_score</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>cluster_alias</th>\n",
       "      <th>cell_in_wmb_study</th>\n",
       "      <th>wmb_cluster_alias</th>\n",
       "      <th>library_label</th>\n",
       "      <th>...</th>\n",
       "      <th>Myoc</th>\n",
       "      <th>Sfrp1</th>\n",
       "      <th>Dscaml1</th>\n",
       "      <th>Sox4</th>\n",
       "      <th>Adgrv1</th>\n",
       "      <th>Siah3</th>\n",
       "      <th>Gpc5</th>\n",
       "      <th>Ddn</th>\n",
       "      <th>Aqp4</th>\n",
       "      <th>Emx2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCCTGTTGTGAATTAG-135_B01</th>\n",
       "      <td>GCCTGTTGTGAATTAG</td>\n",
       "      <td>6777</td>\n",
       "      <td>37834.0</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>-1.185296</td>\n",
       "      <td>1.881103</td>\n",
       "      <td>278</td>\n",
       "      <td>True</td>\n",
       "      <td>1079</td>\n",
       "      <td>L8TX_190716_01_D07</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCGAAAGTGAAGCGT-761_A04</th>\n",
       "      <td>TCCGAAAGTGAAGCGT</td>\n",
       "      <td>3396</td>\n",
       "      <td>9585.0</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>14.007499</td>\n",
       "      <td>-0.220336</td>\n",
       "      <td>817</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_210805_01_H01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTACCATGTCGTGGTC-327_A06</th>\n",
       "      <td>TTACCATGTCGTGGTC</td>\n",
       "      <td>4294</td>\n",
       "      <td>10527.0</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>6.550781</td>\n",
       "      <td>-1.378886</td>\n",
       "      <td>804</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_200813_01_H10</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACGGGTCGTACGAGCA-385_D06</th>\n",
       "      <td>ACGGGTCGTACGAGCA</td>\n",
       "      <td>2943</td>\n",
       "      <td>6957.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.650183</td>\n",
       "      <td>0.289086</td>\n",
       "      <td>817</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_201008_01_A12</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTAGGATCTGTCCCA-301_B04</th>\n",
       "      <td>CTTAGGATCTGTCCCA</td>\n",
       "      <td>7907</td>\n",
       "      <td>47406.0</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>6.583878</td>\n",
       "      <td>-10.639808</td>\n",
       "      <td>152</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_200723_01_B10</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGACAAAGAGCCCAA-337_B06</th>\n",
       "      <td>AAGACAAAGAGCCCAA</td>\n",
       "      <td>1767</td>\n",
       "      <td>2909.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-2.582976</td>\n",
       "      <td>17.539478</td>\n",
       "      <td>833</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_200820_01_E01</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTACAGTCCAGCCTT-314_B01</th>\n",
       "      <td>TGTACAGTCCAGCCTT</td>\n",
       "      <td>1640</td>\n",
       "      <td>2716.0</td>\n",
       "      <td>0.012658</td>\n",
       "      <td>-4.430760</td>\n",
       "      <td>17.868887</td>\n",
       "      <td>835</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_200810_01_A08</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACTTCGCACAGCTTA-324_D03</th>\n",
       "      <td>CACTTCGCACAGCTTA</td>\n",
       "      <td>7643</td>\n",
       "      <td>40484.0</td>\n",
       "      <td>0.130000</td>\n",
       "      <td>-11.713069</td>\n",
       "      <td>1.782393</td>\n",
       "      <td>242</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_200813_01_F08</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCATTGCACTAACGT-696.1_B03</th>\n",
       "      <td>TTCATTGCACTAACGT</td>\n",
       "      <td>8523</td>\n",
       "      <td>46961.0</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>-13.516253</td>\n",
       "      <td>8.201415</td>\n",
       "      <td>555</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>L8TX_210624_01_H03</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATCGATGTCCGTAGTA-409_B04</th>\n",
       "      <td>ATCGATGTCCGTAGTA</td>\n",
       "      <td>7975</td>\n",
       "      <td>43591.0</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>-7.976950</td>\n",
       "      <td>-3.573590</td>\n",
       "      <td>127</td>\n",
       "      <td>True</td>\n",
       "      <td>299</td>\n",
       "      <td>L8TX_201029_01_A11</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1162565 rows × 71 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                cell_barcode  gene_count  umi_count  \\\n",
       "cell_label                                                            \n",
       "GCCTGTTGTGAATTAG-135_B01    GCCTGTTGTGAATTAG        6777    37834.0   \n",
       "TCCGAAAGTGAAGCGT-761_A04    TCCGAAAGTGAAGCGT        3396     9585.0   \n",
       "TTACCATGTCGTGGTC-327_A06    TTACCATGTCGTGGTC        4294    10527.0   \n",
       "ACGGGTCGTACGAGCA-385_D06    ACGGGTCGTACGAGCA        2943     6957.0   \n",
       "CTTAGGATCTGTCCCA-301_B04    CTTAGGATCTGTCCCA        7907    47406.0   \n",
       "...                                      ...         ...        ...   \n",
       "AAGACAAAGAGCCCAA-337_B06    AAGACAAAGAGCCCAA        1767     2909.0   \n",
       "TGTACAGTCCAGCCTT-314_B01    TGTACAGTCCAGCCTT        1640     2716.0   \n",
       "CACTTCGCACAGCTTA-324_D03    CACTTCGCACAGCTTA        7643    40484.0   \n",
       "TTCATTGCACTAACGT-696.1_B03  TTCATTGCACTAACGT        8523    46961.0   \n",
       "ATCGATGTCCGTAGTA-409_B04    ATCGATGTCCGTAGTA        7975    43591.0   \n",
       "\n",
       "                            doublet_score          x          y  \\\n",
       "cell_label                                                        \n",
       "GCCTGTTGTGAATTAG-135_B01         0.200000  -1.185296   1.881103   \n",
       "TCCGAAAGTGAAGCGT-761_A04         0.030303  14.007499  -0.220336   \n",
       "TTACCATGTCGTGGTC-327_A06         0.020000   6.550781  -1.378886   \n",
       "ACGGGTCGTACGAGCA-385_D06         0.000000  13.650183   0.289086   \n",
       "CTTAGGATCTGTCCCA-301_B04         0.037037   6.583878 -10.639808   \n",
       "...                                   ...        ...        ...   \n",
       "AAGACAAAGAGCCCAA-337_B06         0.000000  -2.582976  17.539478   \n",
       "TGTACAGTCCAGCCTT-314_B01         0.012658  -4.430760  17.868887   \n",
       "CACTTCGCACAGCTTA-324_D03         0.130000 -11.713069   1.782393   \n",
       "TTCATTGCACTAACGT-696.1_B03       0.200000 -13.516253   8.201415   \n",
       "ATCGATGTCCGTAGTA-409_B04         0.125000  -7.976950  -3.573590   \n",
       "\n",
       "                            cluster_alias  cell_in_wmb_study  \\\n",
       "cell_label                                                     \n",
       "GCCTGTTGTGAATTAG-135_B01              278               True   \n",
       "TCCGAAAGTGAAGCGT-761_A04              817              False   \n",
       "TTACCATGTCGTGGTC-327_A06              804              False   \n",
       "ACGGGTCGTACGAGCA-385_D06              817              False   \n",
       "CTTAGGATCTGTCCCA-301_B04              152              False   \n",
       "...                                   ...                ...   \n",
       "AAGACAAAGAGCCCAA-337_B06              833              False   \n",
       "TGTACAGTCCAGCCTT-314_B01              835              False   \n",
       "CACTTCGCACAGCTTA-324_D03              242              False   \n",
       "TTCATTGCACTAACGT-696.1_B03            555              False   \n",
       "ATCGATGTCCGTAGTA-409_B04              127               True   \n",
       "\n",
       "                            wmb_cluster_alias       library_label  ...  Myoc  \\\n",
       "cell_label                                                         ...         \n",
       "GCCTGTTGTGAATTAG-135_B01                 1079  L8TX_190716_01_D07  ...   NaN   \n",
       "TCCGAAAGTGAAGCGT-761_A04                 <NA>  L8TX_210805_01_H01  ...   NaN   \n",
       "TTACCATGTCGTGGTC-327_A06                 <NA>  L8TX_200813_01_H10  ...   NaN   \n",
       "ACGGGTCGTACGAGCA-385_D06                 <NA>  L8TX_201008_01_A12  ...   NaN   \n",
       "CTTAGGATCTGTCCCA-301_B04                 <NA>  L8TX_200723_01_B10  ...   NaN   \n",
       "...                                       ...                 ...  ...   ...   \n",
       "AAGACAAAGAGCCCAA-337_B06                 <NA>  L8TX_200820_01_E01  ...   NaN   \n",
       "TGTACAGTCCAGCCTT-314_B01                 <NA>  L8TX_200810_01_A08  ...   NaN   \n",
       "CACTTCGCACAGCTTA-324_D03                 <NA>  L8TX_200813_01_F08  ...   NaN   \n",
       "TTCATTGCACTAACGT-696.1_B03               <NA>  L8TX_210624_01_H03  ...   NaN   \n",
       "ATCGATGTCCGTAGTA-409_B04                  299  L8TX_201029_01_A11  ...   NaN   \n",
       "\n",
       "                           Sfrp1 Dscaml1 Sox4 Adgrv1 Siah3  Gpc5  Ddn Aqp4  \\\n",
       "cell_label                                                                   \n",
       "GCCTGTTGTGAATTAG-135_B01     NaN     NaN  NaN    NaN   NaN   NaN  NaN  NaN   \n",
       "TCCGAAAGTGAAGCGT-761_A04     NaN     NaN  NaN    NaN   NaN   NaN  NaN  NaN   \n",
       "TTACCATGTCGTGGTC-327_A06     NaN     NaN  NaN    NaN   NaN   NaN  NaN  NaN   \n",
       "ACGGGTCGTACGAGCA-385_D06     NaN     NaN  NaN    NaN   NaN   NaN  NaN  NaN   \n",
       "CTTAGGATCTGTCCCA-301_B04     NaN     NaN  NaN    NaN   NaN   NaN  NaN  NaN   \n",
       "...                          ...     ...  ...    ...   ...   ...  ...  ...   \n",
       "AAGACAAAGAGCCCAA-337_B06     NaN     NaN  NaN    NaN   NaN   NaN  NaN  NaN   \n",
       "TGTACAGTCCAGCCTT-314_B01     NaN     NaN  NaN    NaN   NaN   NaN  NaN  NaN   \n",
       "CACTTCGCACAGCTTA-324_D03     NaN     NaN  NaN    NaN   NaN   NaN  NaN  NaN   \n",
       "TTCATTGCACTAACGT-696.1_B03   NaN     NaN  NaN    NaN   NaN   NaN  NaN  NaN   \n",
       "ATCGATGTCCGTAGTA-409_B04     NaN     NaN  NaN    NaN   NaN   NaN  NaN  NaN   \n",
       "\n",
       "                           Emx2  \n",
       "cell_label                       \n",
       "GCCTGTTGTGAATTAG-135_B01    NaN  \n",
       "TCCGAAAGTGAAGCGT-761_A04    NaN  \n",
       "TTACCATGTCGTGGTC-327_A06    NaN  \n",
       "ACGGGTCGTACGAGCA-385_D06    NaN  \n",
       "CTTAGGATCTGTCCCA-301_B04    NaN  \n",
       "...                         ...  \n",
       "AAGACAAAGAGCCCAA-337_B06    NaN  \n",
       "TGTACAGTCCAGCCTT-314_B01    NaN  \n",
       "CACTTCGCACAGCTTA-324_D03    NaN  \n",
       "TTCATTGCACTAACGT-696.1_B03  NaN  \n",
       "ATCGATGTCCGTAGTA-409_B04    NaN  \n",
       "\n",
       "[1162565 rows x 71 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cells_with_genes = cell_extended.join(gene_data)\n",
    "cells_with_genes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:allen] *",
   "language": "python",
   "name": "conda-env-allen-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
